---
title: "kmer: an R package for fast alignment-free clustering of biological sequences"
author: "Shaun Wilkinson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: kmer.bib
csl: bioinformatics.csl
vignette: >
  %\VignetteIndexEntry{Introduction to the kmer package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE, message = FALSE, warning = FALSE}
#knitr::opts_chunk$set(out.width='750px', dpi=200)
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```

--------------------------------------------------------------------------------
## Abstract
**Summary:** The **kmer** package contains functions for rapidly 
computing distance matrices and clustering biological sequence datasets 
in R, using fast alignment-free *k*-mer counting.
This package has a range of applications including 
OTU clustering, classification trees, and phylogenetic 
analysis of very large biological sequence datasets.  
**Availability and Implementation:** The **kmer** package is 
released underthe GPL-3 license, and is available for download from 
CRAN <https://CRAN.R-project.org/package=kmer> and github 
<https://github.com/shaunpwilkinson/kmer>.  
**Contact:** shaunpwilkinson\@gmail.com


## Introduction
As biological sequence datasets continue to increase in size due to 
high-throughput sequencing, traditional clustering 
methods based on multiple sequence alignments, pairwise
distance matrices, and agglomerative (bottom-up) tree-building methods
can become computationally infeasible.
Alternative k-mer counting methods involve enumerating all 
*k*-letter words in a sequence through a sliding window of length *k*.
The $n \times 4^k$ matrix of *k*-mer counts 
(where $n$ is the number of sequences)
can then be used in place of a multiple sequence alignment to 
calculate distances and/or build a phylogenetic tree. 
Here, we introduce **kmer**, an R package for clustering large 
sequence datasets using fast alignment-free *k*-mer counting. 
This can be achieved with or without a multiple sequence alignment, 
and with or without a matrix of pairwise distances.
These functions are detailed below with examples of their utility.


## Distance matrix computation
The function `kcount` is used to count 
*k*-mers within a sequence or set of sequences, 
by sliding a window of length *k* along each sequence and
counting the number of times each *k*-mer appears 
(for example, the $4^3 = 64$ possible DNA 3-mers: AAA, AAC, AAG, ..., TTT).
The `kdistance` function computes a fast, 
alignment-free distance matrix, using 
*k*-mer counts to derive the pairwise distances. 
The default distance metric used by `kdistance` is the 
*k*-mer (*k*-tuple) distance measure outlined in Edgar [-@Edgar2004c].
For two DNA sequences $a$ and $b$, the fractional common *k*-mer count over the 
$4^k$ possible words of length $k$ is calculated as:
$$F  = \sum\limits_{\tau}\frac{min (n_a(\tau), n_b (\tau))}{min (L_a , L_b ) - k + 1} \tag{1}$$

where $\tau$ represents each possible *k*-mer, 
$n_a(\tau)$ and $n_b(\tau)$ 
are the number of times $\tau$ appears in each sequence, 
$k$ is the *k*-mer length and $L$ is the sequence length. 
The pairwise distance between $a$ and $b$ is then calculated as: 

$$d = \frac{log(0.1 + F) - log(1.1)}{log(0.1)} \tag{2}$$

For $n$ sequences, the `kdistance` operation has time and memory 
complexity $O(n^2)$ and thus can become computationally infeasible 
when the sequence set is large (i.e. more than ~ 10,000 sequences).
As such, the **kmer** package also offers the function `mbed`, that 
only computes the distances from each sequence to a smaller (or equal) 
sized subset of 'seed' sequences [@Blackshields2010]. 
The default behavior of the `kdistance` function is to select 
$t = (log_2n)^2$ seeds by clustering the sequences 
(*k*-means algorithm with $k = t$), and selecting one
representative sequence from each cluster. 
 
DNA and amino acid sequences can be passed to `kcount`, `kdistance` and `mbed` 
either as a list of non-aligned sequences or a matrix of aligned sequences, 
preferably in either the "DNAbin" or "AAbin" raw-byte format (see the 
**ape** package documentation for more information on these S3 classes). 
Character sequences are supported; however ambiguity 
codes may not be recognized or treated appropriately, since raw ambiguities 
are counted according to their underlying residue frequencies
(e.g. the 5-mer "ACRGT" would contribute 0.5 to the tally for "ACAGT" 
and 0.5 to that of "ACGGT"). This excludes the ambiguity code "N", which 
is ignored.


#### Example 1: Compute k-mer distance matrices for the woodmouse dataset
The R package **ape** [@Paradis2004] contains a dataset of 15 aligned mitochondrial 
cytochrome *b* gene DNA sequences from the woodmouse *Apodemus sylvaticus*, 
originally published in Michaux et al. [-@Michaux2003].
While the **kmer** distance functions do not require sequences to 
be aligned, this example will enable us to compare *k*-mer distances
with the traditional alignment-dependent distances produced by 
`ape::dist.dna`. First, load the dataset as follows:

```{r}
library(ape)
data(woodmouse)
## view the first few rows and columns 
as.character.DNAbin(woodmouse[1:5, 1:5])
```


This is a semi-global ('glocal') alignment featuring some 
incomplete sequences, with unknown characters represented by 
the ambiguity code "n" (e.g. No305). 
To avoid artificially inflating the distances between these partial 
sequences and the others, we first trim the gappy ends by subsetting 
the global alignment (note that the **ape** function `dist.dna` 
also removes columns with ambiguity codes prior to distance computation 
by default).

```{r}
woodmouse <- woodmouse[, apply(woodmouse, 2, function(v) !any(v == 0xf0))]
```

The following code first computes the full $n \times n$ distance matrix, 
and then the embedded distances of each sequence to three randomly selected 
seed sequences. 
In both cases the *k*-mer size is set to 5.

```{r}
### Compute the full distance matrix and print the first few rows and columns
library(kmer)
woodmouse.kdist <- kdistance(woodmouse, k = 5)
print(as.matrix(woodmouse.kdist)[1:7, 1:7], digits = 2)

### Compute and print the embedded distance matrix
set.seed(999)
seeds <- sample(1:15, size = 3)
woodmouse.mbed <- mbed(woodmouse, seeds = seeds, k = 5)
### remove the attributes for printing by subsetting the distance matrix
print(woodmouse.mbed[,], digits = 2)
```


##Clustering
To avoid excessive time and memory use when building large trees 
(e.g. *n* > 10,000), the **kmer** package features the 
function `cluster` for divisive tree building, 
free of both alignment and distance matrix computation. 
This function first generates a matrix of *k*-mer counts, 
and then recursively partitions the matrix row-wise using successive 
k-means clustering (*k* = 2). 
While this method may not necessarily 
reconstruct sufficiently accurate phylogenetic trees for taxonomic 
purposes, it offers a fast and efficient means of producing large trees 
for a variety of other applications. 
These include tree-based sequence weighting (e.g. Gerstein et al. [-@Gerstein1994]), 
guide trees for progressive multiple sequence alignment (e.g. Sievers 
et al. [-@Sievers2011]), and other recursive operations such as
classification tree learning.


#### Example 2: Top-down clustering of the woodmouse data 
In this example we will compare the fast top-down 
method with standard neighbor-joining 
trees constructed from full distance matrices computed using both the 
alignment-free *k*-mer distance and the Kimura [-@Kimura1980] 
distance as featured in the **ape** package examples.

```{r, echo = FALSE}
op <- par(no.readonly = TRUE)
```

```{r, out.width='700px', out.height='550px', dpi=500}
## set out plotting panes
par(mfrow = c(1, 3), mar = c(1, 2, 3, 3), cex = 0.3)

## (1) neighbor joining tree with Kimura 1980 distance
### compute the n x n K80 distance matrix 
woodmouse.dist <- ape::dist.dna(woodmouse, model = "K80") 
### build the neighbor-joining tree
tree1.phylo <- ape::nj(woodmouse.dist)
### export as Newick text
tree1.newick <- ape::write.tree(tree1.phylo)
### import as "dendrogram" object
tree1.dendro <- phylogram::read.dendrogram(text = tree1.newick)
### sort nodes by size
tree1.dendro <- phylogram::ladder(tree1.dendro)
### plot the nj tree
plot(tree1.dendro, horiz = TRUE, yaxt = "n", 
     main = "Neighbor-joining tree with\nK80 distance matrix")

## (2) neighbor joining tree with k-mer distance
### compute the n x n k-mer distance matrix 
woodmouse.kdist <- kdistance(woodmouse, k = 5) 
### build the neighbor-joining tree
tree2.phylo <- ape::nj(woodmouse.kdist)
### export as Newick text
tree2.newick <- ape::write.tree(tree2.phylo)
### import as "dendrogram" object
tree2.dendro <- phylogram::read.dendrogram(text = tree2.newick)
### sort nodes by size
tree2.dendro <- phylogram::ladder(tree2.dendro)
### plot the nj tree
plot(tree2.dendro, horiz = TRUE, yaxt = "n", 
     main = "Neighbor-joining tree with\nk-mer distance matrix (k=5)")

## (3) topdown tree without distance matrix
set.seed(999)
tree3 <- cluster(woodmouse, k = 5, nstart = 20)
## sort nodes by size
tree3 <- phylogram::ladder(tree3)
### plot the topdown tree
plot(tree3, horiz = TRUE, yaxt = "n", 
     main = "Top-down tree without\ndistance matrix (k=5)")
```

```{r, echo = FALSE}
## reset plotting parameters
par(op)
```

**Figure 1:** Comparison of clustering methods for woodmouse sequences 
The top-down and neighbor-joining trees show relatively congruent fine-scale 
topologies, regardless of the distance measure and method used.
However, for large sequence sets, the top-down (divisive) method builds trees 
orders of magnitude faster than 
traditional alignment and distance matrix-based methods, since computation 
time and memory use increasing linearly rather than quadratically with *n*.


This package also features the function `otu` for rapid clustering of sequences
into operational taxonomic units based on a genetic distance (k-mer distance)
threshold. This function performs a similar operation to `cluster` in that it
recursively partitions a k-mer count matrix to assign sequences to groups. 
The top-down splitting continues until the farthest k-mer distance
in every cluster is higher than a given threshold value (0.97 by default).
Rather than returning a dendrogram, `otu` returns a named integer vector of 
cluster membership, with asterisks indicating the representative sequences
within each cluster. 


####Example 3: OTU clustering of the woodmouse data
In this final example, the woodmouse dataset is clustered into operational 
taxonomic units with a threshold of 0.97 (k-mer distance with k = 5) and
with 20 random starts per k-means split (recommended for improved accuracy). 
```{r}
set.seed(999)
woodmouse.OTUs <- otu(woodmouse, k = 5, threshold = 0.97, nstart = 20)
woodmouse.OTUs
```
The function outputs a named integer vector of OTU membership, with
asterisks indicating the representative sequence from each cluster
(i.e. the most "central" sequence). 
In this case, three distinct OTUs were found, with No305 and N01114S 
forming one cluster (3), No0909S, No0912S, No1103S, No1007S and No1208S
forming another (2) and the remainder falling into cluster 1 in
concordance with the topology of Figure 1c. 


##Concluding remarks
The **kmer** package is released underthe GPL-3 license, and is free to distribute
under certain conditions; however it comes with no warranty. Please direct bug 
reports to the GitHub issues page at <http://github.com/shaunpwilkinson/kmer/issues>.
Any feedback is greatly appreciated.


## Acknowledgements 
This software was developed with funding from a Rutherford Foundation Postdoctoral 
Research Fellowship from the Royal Society of New Zealand.

## References 
